tests/__init__.py

```diff
@@ -260,6 +268,61 @@ def getresponse(self):
         raise http_client.BadStatusLine("")
 
 
+def _get_free_port():
+    s = socket.socket(socket.AF_INET, type=socket.SOCK_STREAM)
+    s.bind(('localhost', 0))
+    address, port = s.getsockname()
+    s.close()
+    return port
+
+
+class _MockServerRequestHandler(BaseHTTPRequestHandler):
+    """Server request handler which always returns 200 and saves client cert info."""
+    def do_GET(self):
+        # save client cert
+        self.server.last_client_cert = self.connection.getpeercert()
+        # Process an HTTP GET request and return a response with an HTTP 200 status.
+        self.send_response(200)
+        self.end_headers()
+        return
+
+
+class MockHttpServer():
```

Comment:
There already exists code that does same function, right below this class.

Reply:
I respectfully disagree with that statement.
The code below doesn't support SSL and has some hacky "with" implementation / object life time management. It's also impossible to get back client cert used by the server.

Reply:
Yeah, thanks for criticism, I very much agree that code turned out less simple than it should.

So if you disagree here, please wait for me to add https/cert support in existing server stub. There is no good reason to have two testing server implementations.